refactor(compiler): Generate `attribute` and `attributeInterpolate` instructions in template pipeline

dylhunn opened the pull request:
This commit adds the ability to generate attribute instructions as a result of property bindings such as `[attr.foo]='bar'` or `attr.foo='{{bar}}'`. "Singleton" interpolations, such as the previous example, will also be transformed into a simple `attribute` instruction.

(This is based on a previous commit, in which @mmalerba and I recently refactored of the way property bindings are injested.)

eneajaho commented on packages/compiler/src/template/pipeline/ir/src/enums.ts:
```diff
@@ -102,6 +102,11 @@ export enum OpKind {
   Attribute,
+
+  /**
+   * An operation interpolate text into an attribute binding.
```
```suggestion
   * An operation to interpolate text into an attribute binding.
```

mmalerba commented on packages/compiler/src/template/pipeline/ir/src/ops/update.ts:
```diff
@@ -209,6 +205,56 @@ export function createInterpolatePropertyOp(
+
+  /**
+   * The `XrefId` of the template-like element the attribute will belong to.
+   * TODO: This is not currently used; should it be?
```
I see in this PR you haven't hooked it into the attribute_extraction phase yet, but I think once you do this will indeed be used.

mmalerba commented on packages/compiler/src/template/pipeline/ir/src/ops/update.ts:
```diff
@@ -209,6 +205,56 @@ export function createInterpolatePropertyOp(
+  /**
+   * The kind of attribute.
+   */
+  attributeKind: ElementAttributeKind;
```
For style and class bindings, I just [asserted that the kind was binding](https://github.com/angular/angular/pull/50489/files#diff-b1e4c95e3f14b52160cbf995deb6a31f915584b15e5be61ba1ef5dbf7c0fdea5R266). I'm fairly certain that trying to put it on a template is [always an error](https://stackblitz.com/edit/stackblitz-starters-2wgaxy?file=src%2Fmain.ts) anyways. If there turns out to be some edge case we're not thinking of the assert error will make it obvious what happened

dylhunn replied:
Yes -- this makes sense. What would it even mean semantically to set an attribute on an `ng-template`? 🤔
I'm dreading running a TGP and finding out whether people rely on stuff like this.

dylhunn replied:
(I retained the field, but added an explicit sanity check in ingest.)

mmalerba replied:
I assume you're retaining it for structural similarity to the prop op types? You could probably tighten up the type to just `ElementAttributeKind.Binding` if you want

dylhunn replied:
Done!

mmalerba commented on packages/compiler/src/template/pipeline/src/ingest.ts:
```diff
@@ -273,6 +282,13 @@ function ingestPropertyBinding(
         view.update.push(ir.createPropertyOp(xref, bindingKind, name, convertAst(value, view.tpl)));
         break;
+      case e.BindingType.Attribute:
+        debugger;
```
stray debugger

dylhunn commented on packages/compiler/src/template/pipeline/src/instruction.ts:
```diff
@@ -288,6 +316,31 @@ const PROPERTY_INTERPOLATE_CONFIG: VariadicInstructionConfig = {
+const ATTRIBUTE_INTERPOLATE_CONFIG: VariadicInstructionConfig = {
+  constant: [
+    // TODO: Why is this identifier missing?
+    // Identifiers.attributeInterpolate,
```
Awesome, makes sense. Thanks :)

dylhunn commented on packages/compiler/src/template/pipeline/src/instruction.ts:
```diff
@@ -212,6 +216,30 @@ export function propertyInterpolate(
   return callVariadicInstruction(PROPERTY_INTERPOLATE_CONFIG, [o.literal(name)], interpolationArgs);
 }
 
+// TODO: Refactor this into a single function with propertyInterpolate.
```
Sounds good. I'll refactor this out after your PR lands, and reuse your logic.

dylhunn replied:
Done.

mmalerba commented on packages/compiler/src/template/pipeline/src/phases/attribute_extraction.ts:
```diff
@@ -29,6 +29,23 @@ function lookupElement(
 export function phaseAttributeExtraction(cpl: ComponentCompilation, compatibility: boolean): void {
   for (const [_, view] of cpl.views) {
     populateElementAttributes(view, compatibility);
+    collapseSingletonInterpolations(view);
```
I think you can skip all of this if you just put `Identifiers.attribute` in the 0 slot for `ATTRIBUTE_INTERPOLATE_CONFIG`, that's what I did for styleProp, styleMap, etc. or did that feel too hacky to you?

dylhunn replied:
Not too hacky. I copied your approach.

dylhunn committed:
refactor(compiler): Generate `attribute` and `attributeInterpolate` instructions in template pipeline

This commit adds the ability to generate attribute instructions as a result of property bindings such as `[attr.foo]='bar'` or `attr.foo='{{bar}}'`. "Singleton" interpolations, such as the previous example, will also be transformed into a simple `attribute` instruction.

dylhunn commented:
I have rebased this on top of the latest style/property changes. @alxhub, please have a look when you have a moment! This is pretty straightforward, so I want to get it in before the next round of property binding refactors.

dylhunn commented:
caretaker: presubmit and windows CI jobs are flakes

dylhunn commented:
This PR was merged into the repository by commit 875851776c2f1a688554e91b8f5352984a71d16b.
